c/src/space.rs

```diff
@@ -193,77 +184,109 @@ pub extern "C" fn space_observer_free(observer: *mut space_observer_t) {
 ///
 /// replace
-///   \arg payload \c is the pointer to the space's payload
+///   \arg params \c is the pointer to the space's params
 ///   \arg from \c is the atom to replace in the space.  This function should NOT take ownership of the `from` atom.
 ///   \arg to \c is the atom to replace it with.  This function SHOULD take ownership of the `to` atom.
 ///   Returns `true` if the atom was replaced, otherwise returns `false`
 ///
 /// free_payload
 ///   \arg payload \c is the pointer to the space's payload
 ///   NOTE: This function is responsible for freeing the payload buffer, as well as any other objects
-///   and resources stored by the space.  This includes `atom_t` objects, `space_observer_t` objects,
-///   as well as any other buffers stored in the space
+///   and resources stored by the space.  This includes `atom_t` objects, as well as any other buffers
+///   stored in the space
 #[repr(C)]
 pub struct space_api_t {
-    register_observer: extern "C" fn(payload: *mut c_void, observer: *mut space_observer_t),
-
-    query: extern "C" fn(payload: *mut c_void, atom: *const atom_t) -> *mut bindings_set_t,
+    query: extern "C" fn(params: *const space_params_t, atom: *const atom_t) -> *mut bindings_set_t,
 
-    subst: Option<extern "C" fn(payload: *mut c_void, pattern: *const atom_t, tmpl: *const atom_t) -> *mut vec_atom_t>,
+    subst: Option<extern "C" fn(params: *const space_params_t, pattern: *const atom_t, tmpl: *const atom_t) -> *mut vec_atom_t>,
 
-    add: extern "C" fn(payload: *mut c_void, atom: *mut atom_t),
+    add: extern "C" fn(params: *const space_params_t, atom: *mut atom_t),
 
-    remove: extern "C" fn(payload: *mut c_void, atom: *const atom_t) -> bool,
+    remove: extern "C" fn(params: *const space_params_t, atom: *const atom_t) -> bool,
 
-    replace: extern "C" fn(payload: *mut c_void, from: *const atom_t, to: *mut atom_t) -> bool,
+    replace: extern "C" fn(params: *const space_params_t, from: *const atom_t, to: *mut atom_t) -> bool,
 
     free_payload: extern "C" fn(payload: *mut c_void),
 }
 
+pub struct space_observer_list_t {
+    observers: RefCell<Vec<Weak<RefCell<dyn SpaceObserver>>>>
+}
+
+/// Notifies all observers of an event
+#[no_mangle]
+pub extern "C" fn space_observer_list_notify_all(list: *const space_observer_list_t, event: *const space_event_t) {
+    let list = unsafe{ &(*list).observers };
+    let event = unsafe{ &(*event).event };
+
+    let mut cleanup = false;
+    for observer in list.borrow().iter() {
+        if let Some(observer) = observer.upgrade() {
+            observer.borrow_mut().notify(event);
+        } else {
+            cleanup = true;
+        }
+    }
+    if cleanup {
+        list.borrow_mut().retain(|w| w.strong_count() > 0);
+    }
+}
+
+/// Data associated with this particular space, including the space's payload and observers
+#[repr(C)]
+pub struct space_params_t {
+    payload: *mut c_void,
+    observers: Box<space_observer_list_t>,
```

Comment:
Why does it need to be boxed?

Reply:
The main issue is that `space_params_t` is `#[repr(C)]` while `space_observer_list_t` isn't.  We need the RefCell, Vec, Weak, the dyn object, and whatever other Rust types to be behind an opaque reference. (opaque to C)

One option is to make the whole of `space_params_t` opaque, but the downside there is that we'd need an accessor function to get the payload, and that accessor potentially couldn't be inlined because it would cross a library boundary.

If the whole implementation were C, I might try and use the structure header for public stuff and the rest of the allocation for internals to eliminate the pointer-chase.  But that is going to complicate the code and I think we should save the work & complexity for the places where we have data indicating it's on a critical path.  (This might be on a critical path, but I haven't seen data yet)

Reply:
Thanks for the explanation Luke, now it is clear. I asked because I was surprised we could use `Box<some-rust-type>` with `cbindings` to pass the pointer into a C layer. I am curious if we could get rid of `atom_into_ptr` just using boxing? But probably it is not true because we cannot pass ownership to the C code using this trick.

Reply:
Right.  `atom_into_ptr` allocates a box, thus transferring ownership to a heap object, and then decomposes the box into a pointer, ensuring it won't be dropped.  (When I talk about transferring ownership it's just in the compiler's internal bookkeeping.  In many cases, the compiler is likely able to actually optimize the atom creation / initialization  together with the code of `atom_into_ptr`, so no memory copy ever needs to happen.)

In the case of space_observer_list_t, it's just not about ownership but also fundamental representation in memory.  We could pack it all together into one allocation, but we'd have to do some things that are a little unnatural in Rust, and IMO it's not worth it here.

This is a tangent, but when we get further along in the process of optimization, I suspect we may want a representation of atom_t that bypasses the Box allocation.

Reply:
Thanks!